cpp/include/cudf/detail/offsets_iterator_factory.cuh

```diff
@@ -0,0 +1,87 @@
+/*
+ * Copyright (c) 2023, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/detail/offsets_iterator.cuh>
+
+namespace cudf {
+namespace detail {
+
+/**
+ * @brief Use this class to create an offsetalator instance.
+ */
+struct offsetalator_factory {
+  /**
+   * @brief A type_dispatcher functor to create an input iterator from an offsets column
+   */
+  struct input_offsetalator_fn {
+    template <typename T,
+              std::enable_if_t<std::is_same_v<T, int32_t> or std::is_same_v<T, int64_t>>* = nullptr>
+    input_offsetalator operator()(column_view const& indices)
+    {
+      return input_offsetalator(indices.data<IndexType>(), indices.type());
+    }
+    template <typename T,
+              typename... Args,
+              std::enable_if_t<not std::is_same_v<T, int32_t> and not std::is_same_v<T, int64_t>>* =
+                nullptr>
+    input_offsetalator operator()(Args&&... args)
+    {
+      CUDF_FAIL("offsets must be int32 or int64 type");
+    }
+  };
+
+  /**
+   * @brief Create an input offsetalator instance from an offsets column
+   */
+  static input_offsetalator make_input_iterator(column_view const& offsets)
+  {
+    return type_dispatcher(offsets.type(), input_offsetalator_fn{}, offsets);
+  }
```

Comment:
As we only have two supported types, it might be better to not invoke `type_dispatcher` and just explicitly instantiate the functor for int32 and int64?

Reply:
That is not our usual pattern. I'd rather the type-dispatcher handle dispatching on the type.

Reply:
On second thought, the dispatch functor does not actually use the resolved type other than to check what it is. And since there are only two types, a runtime check of the types would be cleaner and clearer.

Reply:
Yes, that was my understanding too. The functor needs full type coverage, otherwise type_dispatcher would fail with a compiler error. But I'm not sure what the overhead would be here to invoke this every time we create an instance of this iterator, what do you think?

Reply:
The overhead should be small since there are only 2 types and the compiler is good at optimizing out the other paths.
I created this example: https://godbolt.org/z/44ExEsn3s to show how the `std::sqrt()` function is never actually materialized in the assembly code.

Reply:
Are  you asking how the compiler performs the optimization? I'm not sure how to check for that. 
You can remove the `-O3` option in the godbolt and see the call to `sqrt()` does show up 2 times (for int8 and int16).

Reply:
There does seem to be a compile issue in time and size with this approach.
I need to rethink how the offsetalator is implemented to see if we can take advantage of optimizing for just the two types.